### PR TITLE
Turn it into a proper Python package

### LINKED ISSUE #580: Making MNE-BIDS-Pipeline a proper Python package
It was a conscious design decision **not** to make the pipeline a Python package for the following reasons:

- The code shouldn't be "hidden" from the users, to facilitate looking at and potentially editing the scripts
- It should be possible to have multiple versions of the pipeline installed within the same Python environment, to make working on multiple projects in parallel easier

I believe it's safe to assume that:
- At least since we added support for parallelization and now caching, the complexity of our scripts has exploded and they're not trivial to understand to many users anymore. This is regrettable but apparently unavoidable if we want to have the features that are required to improve pipeline adoption.
- Even before complexity increased to the current level, contributions from folks outside of the "@agramfort bubble" were only marginal.
- It's trivially simple and IMHO good practice to create separate virtual Python environments for every project and analysis you're working on.
- Installation and updating of the pipeline is currently all but user-friendly and I believe a **major** blocker for pipeline adoption.
- The manual steps required to create a custom config file are painful and not well-documented or not documented at all, causing users to do "wrong" things like copying config.py and forgetting to remove all the clutter below that "custom config ends here" horizon, which leads to an infinite recursion when starting the pipeline. 

I had started working on the MNE-BIDS-Pipeline-Runner a while back, attempting to ease this pain; but I never had enough time to bring this project to a usable state:

https://github.com/hoechenberger/mne-bids-pipeline-runner

Ultimately, it was just a workaround for the fact that the pipeline was never available as a package. But I like the idea of providing an interactive prompt that just asks you what to do. 

But with all of the above being said -- I believe it's finally time to:

- Make the pipeline a proper Python package
- Add an entry point to run the pipeline from the command line
- Add an entry point to create a custom / template config. 
- Expose the functionality from `run.py::process()` nicely such that users can invoke the pipeline from their scripts. This bit invaluable to me when it comes to debugging. In fact, every time I wish to debug the pipeline, I create an `mwe.py` inside the pipeline directory, which imports `process` from `run.py`. Only that way can I use the VS Code debugger that I'm used to work with. 

A package could also be installed in "editable" mode, so interested parties could still dump the repository folder wherever they want on their hard drive and edit / look at the code at will. 

I'd like to hear your opinions on the thoughts and proposals above 👍

### PR DESCRIPTION
**Pull request by hoechenberger**
Fixes #580

Very much WIP at this stage, just trying to get the ball rolling.

cc @larsoner 

### Before merging …

- [ ] Changelog has been updated (`docs/source/changes.md`)

### CODE REVIEW

**Comment by hoechenberger** [comment]
@larsoner Now it actually installs and creates an entry-point:

```shell
❯ mne-bids-pipeline
ERROR: Missing required flags: {'config'}
Usage: mne-bids-pipeline <flags>
  optional flags:        --config
  required flags:        --steps | --root_dir | --subject | --session |
                         --task | --run | --interactive | --n_jobs | --debug |
                         --cache
  additional flags are accepted

For detailed information on this command, run:
  mne-bids-pipeline -- --help
```

**Comment by larsoner** [comment]
Nice! Since all the `mne` commands are underscored like `mne browse_raw` here it probably makes sense for consistency to use `mne_bids_pipeline ...`

**Comment by hoechenberger** [comment]
> Nice! Since all the `mne` commands are underscored like `mne browse_raw` here it probably makes sense for consistency to use `mne_bids_pipeline ...`

`mne_bids` uses underscores too, so I suppose you're right. Although I have to admit I do not like the inconsistency between package name and entry-point name! It's confusing … and the underscores are not very UNIX-y 🤔

**Comment by hoechenberger** [comment]
Shall we switch to running all tests via the CLI too?

**Comment by hoechenberger** [comment]
@larsoner Using MNE's `run_subprocess` introduces an additional newline / empty line after each line printed by the subprocess here. Do you know how to fix this?

**Comment by larsoner** [comment]
Why do you need run_subprocess?

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,7 @@ __pycache__/
 .dask-worker-space/
 mprofile_*.dat
 .joblib_cache/
+dist/
+mne_bids_pipeline.egg-info/
+build/
+dist/
```
Comment (hoechenberger) [inline .gitignore:19]:
```suggestion
```

**Comment by hoechenberger** [comment]
Re logging, I've opened a bug upstream: https://github.com/mne-tools/mne-python/issues/11218

**Comment by hoechenberger** [comment]
@larsoner @agramfort 
This should be good to merge once CI comes back green


The changes here allow installation of the pipeline via `pip install .` (editable mode supported as well), and `python run.py ...` can simply be replaced with `mne_bids_pipeline ...`

In followup PRs, we can add features like "please create a template config file" etc.

Once we have done that, I believe it's time for a first release!

File: .github/workflows/run-tests.yml [hunk 2]
```diff
@@ -30,7 +30,7 @@ jobs:
     - name: Lint config.py with flake8
       run: make flake-config
     - name: Install codespell
-      run: pip install codespell
+      run: pip install "codespell @ https://github.com/Freed-Wu/codespell/archive/refs/heads/iss2055.zip" tomli  # fork that supports pyproject.toml
```
Comment (larsoner) [inline .github/workflows/run-tests.yml:33]: -1 on this, let's push for https://github.com/codespell-project/codespell/issues/2055 instead

File: run.py [hunk 3]
```diff
@@ -316,5 +316,9 @@ def process(
         logger.info('Done running 👆', extra=extra)
 
 
-if __name__ == '__main__':
+def main():
     fire.Fire(process)
```
Comment (agramfort) [inline run.py:320]: run.py is still useful?

**Review by agramfort — COMMENTED** [review]
@hoechenberger thanks 

can you see if the doc needs any update?

File: BUILDING.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Building a release
+
+* Tag a new release with `git` if necessary.
+* Create `sdist` distribution:
+
+  ```shell
+  python -m build --sdist
```
Comment (larsoner) [inline BUILDING.md:7]: you should always also create a wheel so that `--only-binary ":all:"` works

**Review by larsoner — COMMENTED** [review]
A remarkably small diff actually!

I would prefer that we change the folder structure to reflect the standard scientific Python folder layout, i.e., nest all Python under a `mne_bids_pipeline/*` rather than at the repo root. The root would then just have `docs`, `mne_bids_pipeline`, and `pyproject.toml`.

**Comment by hoechenberger** [comment]
> I would prefer that we change the folder structure to reflect the standard scientific Python folder layout

I actually actively wanted to avoid this until now, to cause as little disruption as possible to long-term users who expect to use `run.py`! But we can make the change if you think it'd make sense

File: run.py (hunk 3 above)
Comment (hoechenberger) [reply run.py:320]: Yep, it was one of my goals here to get everything working with as little changes as possible!

But see https://github.com/mne-tools/mne-bids-pipeline/pull/611#issuecomment-1265968499

**Comment by larsoner** [comment]
I think *a lot* of the internal mechanics might be simplified if we can make everything private by moving `run.py` to `mne_bids_pipeline/_run.py` etc. I've seen quite a bit of redundant code and gymnastics in all the automation routines that could be simplified by this. Just a few examples:

1. No need for `sys.path` insertion because we can use proper package-relative imports like `from ._scripts import ...` rather than absolute `from scripts import ...` which failed before (for me at least)
2. Code dedup between `config.py` and `run.py`, e.g., related to logging and imports
3. Moving all helper functions out of `config.py` to a new `_utils.py` to make the `config.py` "pure" in the sense that it can contain just all the default config values -- so we can both use it ourselves, *and* copy-paste it to give people a default config (probably while commenting out all lines like in standard Unix config files)
4. Simplification of doc building gymnastics due to interactions between variables (e.g., https://github.com/mne-tools/mne-bids-pipeline/pull/619 would be easier I think)

If we support `python run.py ...` for people rather than making them transition to `mne_bids_pipeline ...` we have to wait on this stuff :(

**Comment by larsoner** [comment]
Okay I'll leave the refactor to a follow-up PR since it's going to be a lot of `git mv` that would be nice to keep separate from this package-related stuff. I'll just push a quick fix for my comments above and then mark for merge!

**Comment by hoechenberger** [comment]
> I think a lot of the internal mechanics might be simplified if we can make everything private by moving run.py to mne_bids_pipeline/_run.py etc.

Yep! That code will become much easier to maintain

**Comment by hoechenberger** [comment]
> Okay I'll leave the refactor to a follow-up PR since it's going to be a lot of `git mv` that would be nice to keep separate from this package-related stuff. I'll just push a quick fix for my comments above and then mark for merge!

Sounds good to me!

File: .github/workflows/run-tests.yml (hunk 2 above)
Comment (larsoner) [reply .github/workflows/run-tests.yml:33]: ... I see why you did this, so you could get rid of `setup.cfg` entirely. Let's keep it then and I'll fix this once the upstream PR is merged

**Comment by larsoner** [comment]
Thanks @hoechenberger !

**Comment by agramfort** [comment]
isn't it just asking people to use mne_bids_pipeline instead of run.py ?

**Comment by hoechenberger** [comment]
@agramfort I don't know the context of your response :) do you mean how we would need to update the docs? We already did do that!

**Comment by agramfort** [comment]
Sorry this email was sent with a delay from my iPad …

**Pull request merged**